test_palindrome.py

```diff
@@ -0,0 +1,151 @@
+        TestCaseNoOptions(expectation="Expected string, got <class 'float'>!", data=0.0),
+    ]
+    for expectation, data in test_cases:
+        with pytest.raises(ValueError) as value_error:
+            is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert str(value_error.value) == expectation, \
+            f"Expected '{expectation}', got '{str(value_error.value)}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_empty_string_is_a_palindrome(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    expectation, data = True, ""
+    result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+    assert result == expectation,\
+        f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+        f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_strict_palindromes_that_are_not_affected_by_options(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    TestCaseNoOptions = namedtuple(
+        typename='TestCaseNoOptions',
+        field_names=["expectation", "data"]
+    )
+    test_cases = [
+        TestCaseNoOptions(expectation=True, data="abba"),  # lower case even length
+        TestCaseNoOptions(expectation=True, data="towot"),  # lower case odd length
+        TestCaseNoOptions(expectation=True, data="ave eva"),  # with white space
+        TestCaseNoOptions(expectation=True, data="ABBA"),  # upper case
+        TestCaseNoOptions(expectation=True, data="AbbA"),  # mixed case
+        TestCaseNoOptions(expectation=True, data="xml:lmx"),  # with punctuation inside
+        TestCaseNoOptions(expectation=True, data="!abba!"),  # with punctuation outside
+    ]
+    for expectation, data in test_cases:
+        result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert result == expectation, \
+            f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_palindromes_option_case_sensitive(whitespace_sensitive: bool, punctuation_sensitive: bool) -> None:
```

Comment:
Zaczyna się robić skomplikowany przypadek :-)

Może tutaj skupilibyśmy się na pozostawieniu `whitespace_sensitive` i `punctuation_sensitive` w wartości domyślnej, a sparametryzowali tylko `case_sensitive`?

```
@pytest.mark.parametrize("case_sensitive", [True, False])
def test_palindromes_option_case_sensitive(case_sensitive: bool) -> None:
    # if case sensitive is True then 'Kajak' is not a palindrome
    expectation = not case_sensitive   
    assert is_palindrome("Kajak") is expectation
```

Co o tym myślisz?

Reply:
Na pewno będzie łatwiej wyczytać co się dzieje jak nie będzie wszystkich tych parametrize od nie testowanych opcji.

Jedynie dla czytelności rozważyłbym propozycje jak zwcześniejszego komentarza (https://github.com/gstopa/kodilla-automated-tests/pull/4#discussion_r840620955):
```
@pytest.mark.parametrize(
    "case_sensitive,expectation",
    [
         (True, False),
         (False, True),
    ],
)
def test_palindromes_option_case_sensitive(case_sensitive: bool, expectation: bool) -> None:
    # if case sensitive is True then 'Kajak' is not a palindrome
    assert is_palindrome("Kajak") is expectation
```

W takiej sytuacji mamy deklarowaną w wejściu do testu oczekiwaną wartość, i nie trzeba się zastanawiać jak się ją wylicza.

Reply:
Tak, to nawet bardziej mi się podoba :+1: